cpu/em_quad_level_frequency_transitions.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+#
+# Copyright: 2018 IBM
+# Author: Shriya Kulkarni <shriyak@linux.vnet.ibm.com>
+#        : Praveen K Pandey <praveen@linux.vnet.ibm.com>>
+
+import time
+import random
+import platform
+from avocado import Test
+from avocado import main
+from avocado.utils import process, distro, cpu
+from avocado import skipIf
+from avocado.utils.software_manager import SoftwareManager
+
+IS_POWER_NV = 'POWER9' not in open('/proc/cpuinfo', 'r').read()
+
+
+class freq_transitions(Test):
+    """
+    To validate quad level frequency transitions.
+    """
+    @skipIf(IS_POWER_NV, "This test only supported on Power9  platform")
+    def setUp(self):
+        """
+        Verify :
+        1. It is Power system and platform is Power NV.
+        2. Cpupower tool is installed.
+        """
+
+        if 'ppc' not in distro.detect().arch:
+            self.cancel("Processor is not ppc64")
+
+        smm = SoftwareManager()
+        detected_distro = distro.detect()
+        self.threshold = int(self.params.get("threshold", default=300000))
+        if 'Ubuntu' in detected_distro.name:
+            deps = ['linux-tools-common', 'linux-tools-%s'
+                    % platform.uname()[2]]
+        elif detected_distro.name == "SuSE":
+            deps = ['cpupower']
+        else:
+            deps = ['kernel-tools']
+        for package in deps:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+
+        self.cpus = cpu.total_cpus_count()
+        self.cpu_num = 0
+        self.max_freq = 0
+        self.quad_dict = {}
+        self.max_freq_dict = {}
+        self.quad_to_cpu_mapping()
+
+    def test(self):
+        """
+        1) Change governor to userspace on all CPUs.
+        2) For each core in the quad, pick a random frequency
+        and set cpu with that frequency.
+        3) Validate that the cpuinfo_cur_freq on any core in
+        the code is set to max(set of frequencies)
+        4) Run the perf command on each cpu to validate frequencies
+        independently.
+        """
+
+        output = process.run("cpupower frequency-set -g userspace", shell=True)
```

Comment:
i think not needed as `process.run ` not going to error out any time

Reply:
It might sometimes throw `CmdError` right?
```
process.run("ls a", shell=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/site-packages/avocado_framework-61.0-py2.7.egg/avocado/utils/process.py", line 1265, in run
    raise CmdError(cmd, sp.result)
avocado.utils.process.CmdError
```

Reply:
sure i meant to say about this specific command `cpupower frequency-set -g userspace` but good suggestion i will come with as sub routine which does process.run and call from all the places